### ISSUE TITLE
Function `reRoll()` is deterministic - it's possible to calculate what figheter's traits we will re-roll

### ISSUE THREAD

**Issue by c4-bot-6** [code-423n4/2024-02-ai-arena-findings#721]
# Lines of code

https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L370-L390

# Vulnerability details

## Impact

Function `reRoll()` is used to rolls a new fighter with random traits.
However, the `dna`, which is used to calculate traits is depended on `msg.sender` address, thus it's deterministic. Addresses in blockchain are deterministic. Before starting the interaction with AI Arena, user may pre-determine multiple of different smart contract addresses before deployment or generate multiple of EOAs - and them use those addresses to pre-calculate `reRoll()` outcome.

This will allow the attacker to choose the best address (`msg.sender`) which he should use to interact with AI Arena. After calling `reRoll()` by using that address - he will receive the desired new fighter (with previously pre-calculated rarity and attributeIndex).

This implies, that re-rolling a new fighter is not random at all and can be pre-calculated.

## Proof of Concept

[File: FighterFarm.sol](https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/FighterFarm.sol#L370-L390)
```
function reRoll(uint8 tokenId, uint8 fighterType) public {
        require(msg.sender == ownerOf(tokenId));
[...]
            uint256 dna = uint256(keccak256(abi.encode(msg.sender, tokenId, numRerolls[tokenId])));
[...]
            fighters[tokenId].physicalAttributes = _aiArenaHelperInstance.createPhysicalAttributes(
                newDna,
                generation[fighterType],
                fighters[tokenId].iconsType,
                fighters[tokenId].dendroidBool
            );
            _tokenURIs[tokenId] = "";
        }
    }    
```

Whenever we call `reRoll()` - the `dna` is calculated as: `uint256(keccak256(abi.encode(msg.sender, tokenId, numRerolls[tokenId])));`.
As we can see, every parameter of keccak256 is known. We know `tokenId`, we know `numRerolls[tokenId]` and we know `msg.sender`.

Then on that `dna`, we call  `_aiArenaHelperInstance.createPhysicalAttributes`. Let's check how it's implemented

[File: AiArenaHelper.sol](https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/AiArenaHelper.sol#L83-L121)
```
function createPhysicalAttributes(
       [...]
                } else {
                    uint256 rarityRank = (dna / attributeToDnaDivisor[attributes[i]]) % 100;
                    uint256 attributeIndex = dnaToIndex(generation, rarityRank, attributes[i]);
[...]
            );
        }
    }
```

The rarity is calculated as `(dna / attributeToDnaDivisor[attributes[i]]) % 100`.
The attributeIndex is calculated as `dnaToIndex(generation, rarityRank, attributes[i])`.
As we can see, when we can control the value of `dna` - we can manipulate the result of rarity and attributeIndex.

There are two possible attack scenarios, which we will describe in the below section.

* Using EOAs

1. Attacker generates multiple of EOAs.
2. Attacker uses each of EOA to calculate the outcome of `reRoll()`. Since every parameter of `dna` calculation is known, he is able to  pre-calculate which EOA will give him desired rarity and attributeIndex after calling `reRoll()`.
3. When attacker decides which EOA is the best for him - he finally uses that EOA to interact with AI Arena.
4. Attacker calls `reRoll()` - he received the fighter with desired (previously calculated) rarity and attributeIndex

* Using smart contracts
1. Attacker creates a smart contract which will interact with AI Arena.
2. `CREATE2` allows to pre-determine contract addresses before deployment.
3. Attacker slightly modifies the contract's code, to pre-determine multiple of different addresses before deployment.
4. He uses those addresses to calculate `dna`.
5. He decides which `dna` gives him the best fighter (with the best rarity and attributeIndex).
6. He chooses that smart contract, deploys it, and starts interact with AI Arena.
7. Deployed smart contract calls `reRoll()` - and receives the fighter with desired (previously calculated) rarity and attributeIndex.

To sum up these attack scenarios - attacker can pre-calculate multiple of addresses and simulate the result of `reRoll()` for each of these addresses. Then he chooses the address which gave him the best fighter and uses that address to interact with Arena AI. He then calls `reRoll()` and receives the desired fighter.

## Tools Used

Manual code review

## Recommended Mitigation Steps

Re-rolls should guarantee full randomness. It should not be possible to pre-calculate fighter rarity and attributeIndex based on the caller's address. It's recommended to add some truly random parameter which will take part in `dna` calculation.

## Assessed type

Other
